pkg/controller/job/job_controller_test.go

```diff
@@ -163,206 +163,323 @@ func TestControllerSyncJob(t *testing.T) {
-		manager, sharedInformerFactory := newControllerFromClient(clientset, controller.NoResyncPeriodFunc)
-		fakePodControl := controller.FakePodControl{Err: tc.podControllerError, CreateLimit: tc.podLimit}
-		manager.podControl = &fakePodControl
-		manager.podStoreSynced = alwaysReady
-		manager.jobStoreSynced = alwaysReady
-		var actual *batch.Job
-		manager.updateHandler = func(job *batch.Job) error {
-			actual = job
-			return nil
-		}
+		t.Run(name, func(t *testing.T) {
+			// job manager setup
+			clientset := clientset.NewForConfigOrDie(&restclient.Config{Host: "", ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
+			manager, sharedInformerFactory := newControllerFromClient(clientset, controller.NoResyncPeriodFunc)
+			fakePodControl := controller.FakePodControl{Err: tc.podControllerError, CreateLimit: tc.podLimit}
+			manager.podControl = &fakePodControl
+			manager.podStoreSynced = alwaysReady
+			manager.jobStoreSynced = alwaysReady
+			var actual *batch.Job
+			manager.updateHandler = func(job *batch.Job) error {
+				actual = job
+				return nil
+			}
 
-		// job & pods setup
-		job := newJob(tc.parallelism, tc.completions, tc.backoffLimit)
-		if tc.deleting {
-			now := metav1.Now()
-			job.DeletionTimestamp = &now
-		}
-		sharedInformerFactory.Batch().V1().Jobs().Informer().GetIndexer().Add(job)
-		podIndexer := sharedInformerFactory.Core().V1().Pods().Informer().GetIndexer()
-		setPodsStatuses(podIndexer, job, tc.pendingPods, tc.activePods, tc.succeededPods, tc.failedPods)
+			// job & pods setup
+			job := newJob(tc.parallelism, tc.completions, tc.backoffLimit)
+			if tc.deleting {
+				now := metav1.Now()
+				job.DeletionTimestamp = &now
+			}
+			sharedInformerFactory.Batch().V1().Jobs().Informer().GetIndexer().Add(job)
+			podIndexer := sharedInformerFactory.Core().V1().Pods().Informer().GetIndexer()
+			setPodsStatuses(podIndexer, job, tc.pendingPods, tc.activePods, tc.succeededPods, tc.failedPods)
 
-		// run
-		forget, err := manager.syncJob(testutil.GetKey(job, t))
+			// run
+			forget, err := manager.syncJob(testutil.GetKey(job, t))
 
-		// We need requeue syncJob task if podController error
-		if tc.podControllerError != nil {
-			if err == nil {
-				t.Errorf("%s: Syncing jobs would return error when podController exception", name)
+			// We need requeue syncJob task if podController error
+			if tc.podControllerError != nil {
+				if err == nil {
+					t.Error("Syncing jobs expected to return error on podControl exception")
+				}
+			} else if tc.failedPods > 0 && tc.expectedCondition == nil {
+				if err == nil {
+					t.Error("Syncing jobs expected to return error when there are new failed pods and Job didn't finish")
+				}
+			} else if tc.podLimit != 0 && fakePodControl.CreateCallCount > tc.podLimit {
```

Comment:
```suggestion
			} else if tc.podLimit != 0 && fakePodControl.CreateCallCount >= tc.podLimit {
```

right?

Reply:
no, it was rather a bug in the existing code :)